File: game theory/predict the winner.cpp
int dp[30][30];
class Solution {
public:
	bool PredictTheWinner(vector<int>& nums) {
		for (int g = 0; g < nums.size(); g++) {
			for (int i = 0, j = g; j < nums.size(); i++, j++) {
				if (g == 0) {
					dp[i][j] = nums[i];
				}
				else if (g == 1) {
					dp[i][j] = max(nums[i], nums[j]);
				}
				else {
					int value = nums[i] + min(dp[i + 2][j], dp[i + 1][j - 1]);
					int value2 = nums[j] + min(dp[i + 1][j - 1], dp[i][j - 2]);
					dp[i][j] = max(value, value2);
				}
			}
		}
		int two = 0;
		for (auto i : nums) two += i;
		int one = dp[0][nums.size() - 1];
		two -= one;
		return one >= two;
	}
};

File: game theory/predict the winner.py
class Solution:
    def solve(self, nums, i, j, chance, dp):
        if (dp[i][j] != -1):
            return dp[i][j]
        if (i > j):
            return 0
        if chance == 0:
            dp[i][j] = max(nums[i] + self.solve(nums, i + 1, j, 1, dp), 
                       nums[j] + self.solve(nums, i, j -1, 1, dp))
        else: 
            dp[i][j] = min(self.solve(nums, i + 1, j, 0, dp), 
                       self.solve(nums, i, j -1, 0, dp))
        return dp[i][j]
    def PredictTheWinner(self, nums: List[int]) -> bool:
        two = sum(nums)
        dp = [[-1 for i in range(len(nums) + 1)]for j in range(len(nums) + 1)]
        one = self.solve(nums, 0, len(nums)-1, 0, dp)
        two -= one
        return one >= two
        
